containers/src/Kokkos_DualView.hpp

```diff
@@ -557,7 +590,27 @@ class DualView : public ViewTraits<DataType, Arg1Type, Arg2Type, Arg3Type> {
     }
   }
 
-  void sync_host() {
+  template <class Device>
+  void sync(const typename std::enable_if<
+                (!std::is_same<typename traits::data_type,
+                               typename traits::non_const_data_type>::value) ||
+                    (std::is_same<Device, int>::value),
+                int>::type& = 0) {
+    sync_impl<Device>(std::false_type{});
+  }
+  template <class Device, class ExecutionSpace>
+  void sync(const ExecutionSpace& exec,
+            const typename std::enable_if<
+                (!std::is_same<typename traits::data_type,
+                               typename traits::non_const_data_type>::value) ||
+                    (std::is_same<Device, int>::value),
+                int>::type& = 0) {
+    sync_impl<Device>(std::false_type{}, exec);
+  }
```

Comment:
Surely you meant
```suggestion
  template <class Device>
  void sync() {
    sync_impl<Device>(std::intergal_constant<bool,
    std::is_same<typename traits::data_type, typename traits::non_const_data_type>::value ||
    std::is_same<Device, int>::value>);
  }
```

Reply:
I don't understand why you mixSFINAE and tag dispatching

Reply:
Honestly, I mean that this should be completely rewritten, with this type trait made a `constexpr static bool` of the class. If you'd like to make this change I can look at it, though I know that as is it wouldn't compile. Also, I don't know what it means for the Device to be an int.

Reply:
Also, wait, your example doesn't pass the execution space, if we're given one?

Reply:
I missed that

Reply:
```C++
  template <class Device, class.. ExecSpaceOrNothing>
  void sync(ExecSpaceOrNothing const& exec_space_or_nothing...) {
    using Tag = std::intergal_constant<bool,
        std::is_same<typename traits::data_type, typename traits::non_const_data_type>::value ||
        std::is_same<Device, int>::value>;
    sync_impl<Device>(Tag{}, exec_space_or_nothing...);
  }

Reply:
and you can try a
```C++
using Tuple = std::conditional_t<sizeof...(ExecSpaceOrNothing)==1, std::tuple< ExecSpaceOrNothing...>, std::tuple<void>>;
static_assert(sizeof...(ExecSpaceOrNothing)==0 || sizeof...(ExecSpaceOrNothing)==1 && Kokkos::is_execution_space<std::tuple_element_t<0,Tuple>>::value, "")

Reply:
or better if you want make Daisy proud
```C++
#ifdef KOKKOS_ENABLE_CXX17
static_assert((... && Kokkos::is_execution_space<ExecSpaceOrNothing>::value));
#endif

Reply:
So, what I worry about here is that in the current version, we have functions with the same signatures I have provided, that is

```c++
template<typename Device> void sync(int)
```

So I am _positive_ that any program using DualView won't see changes. I can argue to myself that your change works: the only thing I can see breaking is if somebody incorrectly actually passed an `int` argument:

```c++
Kokkos::DualView<float*, Space> my_dv;
my_dv.sync<Space>(0);
```

This would be incorrect, but parses and works in the current version. That's just off the top of my head, and this is C++, there's always another corner case. If you think your edit needs to happen, I'll do it, but I worry about doing the subtle and clever thing to save ~30 lines, rather than guaranteeing ourselves the same interface. My recommendation is to have an action item for 4.0 to make this change, but emphasize stability here

Reply:
Ah, thought of another: if you've taken a (member) function pointer to this function, this would be a breaking change. Of course I hope nobody is taking function pointers to these things, but I've hoped for a lot of things...